### PR TITLE
extend terraform and opentofu local deployers to support multiple versions

### PR DESCRIPTION
**Pull request by baixinsui**
Fixed https://github.com/eclipse-xpanse/xpanse/issues/1941

OpenTofu with multiple versions:
image

Terraform with multiple versions:
image

### CODE REVIEW

File: runtime/src/test/java/org/eclipse/xpanse/runtime/modules/deployment/OpenTofuInstallerTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package org.eclipse.xpanse.runtime.modules.deployment;
+
+import static org.eclipse.xpanse.modules.deployment.deployers.opentofu.opentofulocal.OpenTofuInstaller.OPEN_TOFU_VERSION_OUTPUT_PATTERN;
```
Comment (Alice1319) [inline runtime/src/test/java/org/eclipse/xpanse/runtime/modules/deployment/OpenTofuInstallerTest.java:3]: this line of code can be moved below line 11.

File: runtime/src/test/java/org/eclipse/xpanse/runtime/modules/deployment/TerraformInstallerTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package org.eclipse.xpanse.runtime.modules.deployment;
+
+import static org.eclipse.xpanse.modules.deployment.deployers.terraform.terraformlocal.TerraformInstaller.TERRAFORM_VERSION_OUTPUT_PATTERN;
```
Comment (Alice1319) [inline runtime/src/test/java/org/eclipse/xpanse/runtime/modules/deployment/TerraformInstallerTest.java:3]: same to this line.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            try {
+                List<String> versionsInPage =
+                        fetchVersionsFromApi(fetchVersionsApiUrl, page, perPage);
+                if (CollectionUtils.isEmpty(versionsInPage)) {
+                    break;
+                } else {
+                    allVersions.addAll(versionsInPage);
+                    page++;
+                }
+            } catch (IOException e) {
```
Comment (WangLiNaruto) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java:67]: The exception should be handled and the loop terminated

**Comment by Alice1319** [comment]
LGTM

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            case OPEN_TOFU -> OPEN_TOFU_VERSIONS_GITHUB_URL;
+        };
+    }
+
+    private List<String> fetchVersionsFromWebsite(String releasesUrlWithPage) throws IOException {
+        List<String> versions = new ArrayList<>();
+        Document doc = Jsoup.connect(releasesUrlWithPage)
+                .ignoreHttpErrors(true).timeout(5000).get();
+        // version elements with css like <span class="ml-1 wb-break-all">v1.8.0</span>
+        Elements versionElements = doc.select("span.ml-1.wb-break-all");
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java:83]: @baixinsui this looks a bit hacky. We did not use such solution in terraform-boot know?

How about using 'https://api.github.com/repos/hashicorp/terraform/releases' instead? May be we can even use the github Java SDK - https://github-api.kohsuke.org/. 
The releases and tags API of Github works without authentication for public repositories. 
Just an idea.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
+/*
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java:1]: we can move these two classes to another package - deployertools

**Review by swaroopar — CHANGES_REQUESTED** [review]

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java (hunk 4 above)
Comment (baixinsui) [reply modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/deployers/DeployerToolVersionsFetcher.java:83]: I used request the API without the authentication. Due to the limit of 60 times per hour, some tests of the APIs failed.

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java [hunk 6]
```diff
@@ -104,7 +104,7 @@ public class DeployService {
     private ServiceOrderManager serviceOrderManager;
     @Resource
     private ServiceDeploymentStatusChangePolling serviceDeploymentStatusChangePolling;
-    @Resource
+    @Resource(name = "xpanseAsyncTaskExecutor")
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/DeployService.java:107]: can we use the name from here - https://github.com/swaroopar/osc/blob/191b44b81a1113b5675a64260e2b046aac918c36/modules/async/src/main/java/org/eclipse/xpanse/modules/async/TaskConfiguration.java#L21

File: modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/ServiceStateManager.java [hunk 7]
```diff
@@ -57,8 +56,7 @@ public class ServiceStateManager {
     private DatabaseServiceStateManagementTaskStorage taskStorage;
     @Resource
     private DatabaseDeployServiceStorage deployServiceStorage;
-    @Qualifier("xpanseAsyncTaskExecutor")
-    @Resource
+    @Resource(name = "xpanseAsyncTaskExecutor")
```
Comment (swaroopar) [inline modules/deployment/src/main/java/org/eclipse/xpanse/modules/deployment/ServiceStateManager.java:59]: can we use the name from here - https://github.com/swaroopar/osc/blob/191b44b81a1113b5675a64260e2b046aac918c36/modules/async/src/main/java/org/eclipse/xpanse/modules/async/TaskConfiguration.java#L21

**Review by swaroopar — CHANGES_REQUESTED** [review]

**Review by swaroopar — APPROVED** [review]
LGTM

**Pull request merged**